Implement Cicp for ImageBuffer,DynamicImage and provide ColorTransform on moxcms

197g opened the pull request:
The goal here is to bring CICP profiles into `ImageBuffer` / `DynamicImage`. CICP is a perfect fit since, similar to PNG, we only have an rgb-ish pixel canvas in these types (and the sentiment is to have more complex representations, if at all, in separate types).

The first consequence of performance is that we should not generalize over the subpixel type. Handling combinations with sample conversion is only possible when we have an enumeration of possibilities (as in `DynamicImage`) and requires matches with each individual combination handled. Instead, we here convert through `f32` when encountering mismatch to trade a little bit of performance with somewhat terse code.

This uses the internal sealed trait to ensure we only need to handle the `PixelWithColorType` types in the public interface¹. The dispatch that makes this work is a little bit cursed (see comment in `color::private`) but fully hidden away. Also the code is building all tables eagerly for 16 combinations × 3 color depths. This should not be a huge burden in practice. We can patch it later if users want to convert small images infrequently. The whole conversion apparatus allows caching of the `CicpTransformer`, which is also `Clone` and comprised of mainly `Arc<dyn Fn()>` for the individual paths.

Closes: #786
Relates: #1554, #1030, #2493

¹For historical reasons, `Dynamicimage::to_luma_f32` exists without having `ColorType::Luma32F`

fintelia commented on src/color/cicp.rs:
```diff
@@ -0,0 +1,445 @@
+pub struct Cicp {
+    primaries: CicpColorPrimaries,
+    transfer: CicpTransferFunction,
+    matrix: CicpMatrixCoefficients,
+    full_range: CicpVideoFullRangeFlag,
```
We'll have to see whether these two fields end up being needed.

In PNG, only the matrix coefficients representing "RGB" are allowed. And images without the full range [flag set are expected to be extremely rare](https://github.com/w3c/png/issues/312#issuecomment-2327349614) to the point that even browsers aren't planning on supporting them.

awxkee replied:
I believe the fields exist purely for full compatibility. From my view, range and matrix should be handled at the YUV conversion stage and don't make much sense on the colorimetry side.

awxkee reviewed (commented):
Nice to see it here! I don’t think there are any misuses, but I’ve left some thoughts.

awxkee commented on src/color/cicp.rs:
```diff
@@ -0,0 +1,445 @@
+        });
+
+        let gray = {
+            // I'd use record update syntax but there's a private field.
```
AFAIK ```ColorProfile::default``` or ```ColorProfile { foo: bar, ..Default::default() }``` should work.

197g commented on src/color/cicp.rs:
```diff
@@ -0,0 +1,445 @@
+        let tr = CicpTransform::new(Cicp::SRGB, Cicp::DISPLAY_P3).unwrap();
+        let p3 = Rgba::<u8>([255, 0, 0, 255]).to_rgba_with(&tr);
+        assert_eq!(p3.0, [234, 51, 35, 255]);
+    }
```
I've expanded test coverage a little but got stumped. Here's what the test looks like when it succeeds:

```rust
    #[test]
    fn transform_luma() {
        let tr = CicpTransform::new(Cicp::SRGB, Cicp::SRGB).unwrap();
        // _, Y, _ = colour.RGB_to_XYZ(
        //   [1.0, 0.0, 0.0],
        //   colour.RGB_COLOURSPACES['sRGB'],
        //   apply_cctf_decoding=True)
        // colour.RGB_COLOURSPACES['sRGB']._cctf_encoding(Y)*255
        let luma = Rgba::<u8>([255, 0, 0, 255]).to_luma_with(&tr);
        assert_eq!(luma.0, [130u8]); // reference: 127.1021805160301

        let luma = Rgba::<u8>([0, 255, 0, 255]).to_luma_with(&tr);
        assert_eq!(luma.0, [220u8]); // reference: 219.93274897493274

        let luma = Rgba::<u8>([0, 0, 255, 255]).to_luma_with(&tr);
        assert_eq!(luma.0, [70u8]); // reference: 75.962697358369013
    }
```

@awxkee I can see floating point being imprecise but that seems a surprisingly large difference to the reference values. And sometimes too large, some too small. Could you verify my approach? In my implementations in `image-canvas` Ive always achieved accurate-to-rounding test vectors.

awxkee replied:
Your software likely assumes D65 white point colorimetry. moxcms was originally intended to be an ICC converter, and ICC profiles use a fixed D50 white point until ICC MAX (v5) is released. If you want to use D65, you can't use `update_rgb_colorimetry_from_cicp`, because it performs chromatic adaptation to D50.

At least for D50, it matches the tests with reasonable accuracy. I'd even argue which results may be more accurate, but that's not guaranteed.

If you want to perform reference testing, it's better to avoid using `u8`. This is ICC software and is heavily optimized, so `u8`, `u16`, and `f32` (if extended range is disabled) will use LUT tables, which introduce additional quantization.

awxkee replied:
To be clear: all primaries defined in ICC profiles are assumed to be under D50 illumination. As a result, all input is adopted to D50, and from that point the conversion becomes straightforward, since everything is already in a D50 illumination.

This works well for ICC profiles (since ICC profiles should assume D50 white point ). However, it might work not such well if you want different white point or conversion between them. That's not handled out of the box.

197g replied:
While that clears up the difference, technically CICP primaries _also_ specify a reference whitepoint (https://www.itu.int/rec/T-REC-H.273-202407-I/en, Table 2). So it is a bit surprising that the update to colorimetry would not introduce an adaptation matrix accordingly. I suppose for the purpose of further testing, I could set a von-Kries `chromatic_adaptation` by copying over some code from `image-canvas`.

awxkee replied:
> While that clears up the difference, technically CICP primaries also specify a reference whitepoint (https://www.itu.int/rec/T-REC-H.273-202407-I/en, Table 2)

Those are two slightly different things. The primaries being used are correct, but it's how the primaries are defined. The actual transformation takes place under the D50 white point, as required by the ICC standard. The ```ColorProfile.chromatic_adaptation``` should be set to Bradford, since chromatic adaptation is always applied.

You've encountered a small edge case here. It's usually not a big deal for RGB -> RGB conversions, because the adaptation matrix will also be inverted through M^(-1) ( (D65->D50)^(-1) = (D50 -> D65) ), wheareas the final result should still be in D65 sRGB. However, in the case of grayscale, only the D65 -> D50 adaptation is applied, so the result ends up in XYZ values which scaled to D50.

awxkee replied:
[Chrome](https://chromium.googlesource.com/chromium/src/+/lkgr/ui/gfx/color_space.cc) also makes connection on Xyz D50, and I see nothing specific to possibility of Xyz D65. Therefore, I believe it will also consider all gray images as they are D50.

197g replied:
Hm, that still seems odd. The code for luma _does_ copy the chromatic adaptation matrix and `white_point` from the connected rgb color space, hence I assumed that would define the XYZ to working under. So while the pcs is XYZ-D50, should setting the matrix for the gray color space not also result in the inverse being applied when used as a target? Sure, Luma is just extracting the Y component yet that matrix transform is still relevant.

If applying a chad isn't supported for luma then maybe an error should be produced in `create_color_transform`.

awxkee replied:
See https://www.color.org/specification/ICC.1-2022-05.pdf page 101 point F.2. The specification clearly states that `connection = grayTRC[device] ` and `device = grayTRC^(-1) [connection] `.

I understand that it is quite unexpectable, but for ICC software I believe "just following specification" is preferred.

> Sure, Luma is just extracting the Y component yet that matrix transform is still relevant. It should extract Y of XYZ/D65 when a bradford transform XYZ/D65 -> XYZ/D50 is set to define the color spaces relation to the PCS.

I believe it's better to understand ICC chromatic adaptation as "How did you get to D50?", not "What should the output illumination be?'".

197g replied:
Of course, that means we can not expect `sRGB -> grey/D65` to be correct. The encoding of the PCS's Y through greyTRC would be lossy without an inverse whitepoint adaptation, whereas luma->luma is _not_.

197g replied:
There is of course yet another alternative.

> [cicpTag] This tag may be present when the data colour space in the profile header is RGB, YCbCr, or XYZ, and the profile class in the profile header is Input or Display. The tag shall not be present forother data colour spaces or profile classes indicated in the profile header.

What models our intent _best_ is that `Luma` represents the `Y` of an YCbCr data color space. Is that usually encoded as a multiProcessElementsType? Well, and to do that efficiently a `create_masked_transform` is needed that only computes some of the color components (index 0 in this case) of an output data space..

awxkee replied:
I think idea with `greyTRC` which fused with XYZ D50 -> XYZ D65 is good enough.

```multiProcessElementsType``` is quite complicated thing, it require to create not obvious huge LUT which usually in LAB and those are for DeviceLink profiles ( moxcms unfortunately don't support DeviceLink profiles ). I don't think it is the good way.

More common way to link it through `BToA0Tag` and `A0ToBTag` ( for Input Device and Display Device profiles ) but those also require to create much more complicated LUTs.

> Is that usually encoded as a multiProcessElementsType?

Through ICC LUTs, it should be possible to represent almost anything (moxcms can handle any LUTs with up to 16 inks, though I only personally tested up to CMYKOGV, which one I have). However, I think that might be a bit overkill for this case.

More typically ICC LUTs are used for CMYKs, tonemapping, in PDFs some images are not extremely rarely use LAB.

I'll try to make an API and see what the accuracy cost is. However, for common 8-bit transforms I predict that a typical 4096 point curve will be more than enough for good accuracy.

197g replied:
AoBType with B-curves identity, a matrix, and then M curves as the original TRC seems on paper to also have the intended effect for an YCbCr (though I don't see if the adaptation matrix is used in that path, it can as well be fused with the YCbCr matrix manually). So the hinderance would be the 'masked' transform which discards CbCr components in the output and assumes them `0` in the input.

awxkee replied:
That appeared more complicated than expected :)

So my prediction that tone curve is enough to represent tone reproduction curve with XYZ scaling successfully failed. Non linearity is too high, it's not even close.

16-bit lut (b to a) works nice, accuracy loss is low, thus for reference value ```75.94789574522055``` it produce ```75.93887075264664``` what is acceptable. 8-bit LUT makes ```74.02004223358418``` what is not really accurate but not that bad.

> with no CLUT

That's not possible to create transform with different input and output channel count without CLUT. Only through CLUT you can remap PCS XYZ -> 1 output channel and vice verca.

I haven't used 1 channel LUTs before so a few bugs were fixed along the way.

Perhaps I’ll do some API rework in a few days with a minor version update. This will allow to bypass at least for computing D50 white point requirement, since it's still inconvinient if you just make conversion with different white point to do all this stuff. However, it's unclear how to prevent user from saving such profiles, so this probably will be user responsibility not to export "computing profiles".

awxkee replied:
TLDR It’s a nice feature to have, but I probably won’t add it.

Having planar APIs like lcms2 does is actually nice to have.
But if we want to use this in safe Rust, it means no pointer aliasing is available, and that immediately makes things more complicated, since it requires duplicating all the transforms, checks, and other things.
This is a case where C/C++ still wins, as pointer aliasing can be done easily (with known consequences).

This also relates to in-place transforms. I know that many transforms can be done in-place nicely, but this is trade-offs between complexity and safety.

Those two cases will cost about ~20-30K lanes of quite complex code for each, where deinterleaving or allocating additional memory on user side costs 1-3 lanes of very simple code. Therefore, unless I start to work on something with extremely limited resources where such constraints actually matter it's unlikely to happen.

197g replied:
Yes, but having a hard time figuring out if the approach is just hard to get working or if I was thinking about it wrong. So in the meantime I'm considering just leaving out the ability for Luma transforms. (Other than in linear sRGB primary color space which is also what is implemented already within `image` in the pixel convesion, so using that).

It seems like there is necessary complexity; it needs to live in a foundational crate so that moxcms can provide the masked YCbCr interface as mention above with bindings; or it lives in moxcms directly; or it lives in `image` to build a different pipeline for that sample layout (i.e. intermediate allocation). And even the branch does not remove much of that complexity. And I'm all the intricacies of bypassing _some_ conversion _sometimes_ will be hard to test. It also violates one of the principles that the CICP implementation is supposed to follow, that we only use an approach that can be expressed precisely enough to be replicable with other types.

197g replied:
So just to clarify: the `create_gray_to_xyz_samples` and `create_xyz_to_gray_samples` and the utilities to create the right LutMultidimensionalType's make a lot of sense. They are somewhat clearly defined and we can explain them in terms of an actual ICC profile; so that's just using what the specification allows if I see correctly. But the other hacks are rough to use as well. 

So can those kinds of profiles be used to get the right result with little hacks? That route is also encoding the whitepoint adaptation so we can use the D50 PCS as intended. Would it make sense to build those into the create instead of the app binary?

197g commented on src/color/cicp.rs:
```diff
@@ -0,0 +1,445 @@
+
+        let opt = moxcms::TransformOptions::default();
+
+        // TODO: really these should be lazy, eh?
```
With all the different transform tables this is building, we get about `2 << 19` power function executions in total, which is no longer cheap. (16 tables for u8, u16, f32 respectively). The tests take a very noticeable time even with `opt-level = 2` for `moxcms`.

fintelia commented on src/images/buffer.rs:
```diff
@@ -1406,6 +1443,136 @@ where
+    ///
+    /// See [`ImageBuffer::copy_from_color`] if you intend to assign to an existing buffer,
+    /// swapping the argument with `self`.
+    pub fn to_color<IntoType>(
```
Perhaps `apply_color_space` or `apply_color_space_transform`? That would parallel the `apply_orientation` method.

fintelia reviewed (commented):
I like the focus of doing color space conversions on `ImageBuffer` / `DynamicImage`.

How would you feel about making the conversions happen in-place? I think it could significantly simplify the public API. It would remove the potential to convert RGB <-> RGBA or change the sample format at the same time as a color space conversion, but those could always be done in two steps.

The `moxcms` crate doesn't directly support in-place conversions but we could implement it ourselves by iterating over the backing slice with `chunks_mut` and copying each chunk into a stack (or heap) allocated buffer.

197g commented on Cargo.toml:
```diff
@@ -2,10 +2,10 @@
 version = "0.25.6"
 edition = "2021"
-resolver = "2"
+resolver = "3"
```
Good hint, I was under the impression it would choose based on the rust-version of the toolchain compiling the package but it instead uses the `rust-version` field in the crate declaring the dependency. I'll undo this then but am continuing to be surprised. Well it makes sense considering `.lock` file generation.

197g commented:
Allowing two different color types to take place in one operation is a pretty large _feature_.  I'd even like to have different sample types for Input and Output but moxcms does not yet have that capability. The conversion is not *accurate* in all combinations (hence also the use `32F` where it can't use a direct conversion), and (much) more performant directly too—as in allocation-free. I really don't see a good way of teaching or providing that as separate steps. Consequently, I was trying to follow the common designs for BigInt and matrix libraries, with the similarity of requiring some separate in and out buffers for operations. In-place operations are at best a special case.

Following that logic, of course `apply_*` are all in-place methods whereas these methods are not. `to_color` specifically follows the naming of `<[T]>::to_vec` (where `copy_from_color` is conceptually related to `copy_from_slice`). So `DynamicImage::to_color` is still missing but it's target colortype should arguably be a simple argument instead of a type parameter so I'll hold that for a separate PR. There we could also discuss whether to add the special cases of `apply_colorspace` which does _not_

197g commented:
> But I'm still not sure about being able to add/remove an alpha channel at the same time as a color space transform. Having it part of the API while banning RGB <-> Luma conversions adds an unfortunate gotcha. And I'd guess that it wouldn't take meaningfully longer to change color space and alpha channel in two steps versus one?

But converting the bit-depth at the same time is problematic; and addressed by the very same API. 8-bit Luma to 16-bit RGBA makes a lot of sense; as does using another target depth when converting color spaces. That is routine for something like an editor that has an internal surface in lin-rgb RGBA but gets from the outside artbitrary other depths. You can pretty much only do this wrong outside `image` in two-steps, also considering that `DynamicImage` is still marked `#[non_exhaustive]`.

Regarding the claim of consistency, it is the analogue of [copy_from](https://docs.rs/image/latest/image/trait.GenericImage.html#method.copy_from) but correctly modelled with efficiency in mind first, not doing opaque type dispatch that leads to us having to give the problematic advice of writing out the loop pixel-by-pixel. (`SubImage<&mut DynamicImage>` should also get `copy_from_color` but not in this PR).

197g commented:
@fintelia Would you have time for re-review?

The scratch buffer approach works somewhat but is unfortunately slightly more complicated than the sketch. If we would be operating on the most generic `Pixel` there's no good way to 'promote' it to a uniform `Rgbaf<32>`. For Luma we must already do a conversion within that promotion (which depends on the color space) and even for the rgb(a) types the numerics are finicky to express generically. I've thus implemented that transform only for the subset necessary for `PixelWithColor` on `CicpTransform` with the caveat that it won't be available in the public API to transfer its implementation to other kinds of buffers.

197g commented on src/metadata/cicp.rs:
```diff
@@ -0,0 +1,1346 @@
+        let gray = {
+            // I'd use record update syntax but there's a private field.
+            let mut p = rgb.clone();
+            p.color_space = moxcms::DataColorSpace::YCbr;
```
The _compute_ model for these is the YCbCr given by matrix=ChromaDerivedNonConstant though, so in terms of conversion this is correct. But reminder that this is no longer needed, gray conversion does not go through moxcms at this point.

197g commented on src/metadata/cicp.rs:
```diff
@@ -0,0 +1,1346 @@
+        coef: [f32; 3],
+    ) {
+        for (rgba, pix) in input.chunks_exact(4).zip(output.chunks_exact_mut(2)) {
+            let luma = rgba[0] * coef[0] + rgba[1] * coef[1] + rgba[2] * coef[2];
```
We should at least document our wrapper than to note it is meant for cases where only relative accuracy to some known magnitude is required and the precision of fma is merely a side effect.

197g commented on src/images/buffer.rs:
```diff
@@ -1275,6 +1311,7 @@ impl<P: Pixel> ImageBuffer<P, Vec<P::Subpixel>> {
             data: vec![Zero::zero(); size],
             width,
             height,
+            color: Cicp::SRGB_LINEAR.into_rgb(),
```
Probably. With regards to the idea of interpreting the unknown values (`CicpPrimaries::Unspecified = 2`, `CicpTransferFunction::Unspecified = 2`) as sRGB for conversion but not for metadata I'm torn. I think it's fine to assume unwitting _users_ of `new` intend sRGB; and if we do that for compute then arguably we're already saying that the expected perceptual color is sRGB so it's not more wrong to write that to file as is. Users who _do_ know the difference should be able to set the appropriate color space on their buffers before writing them, I'm not very concerned about incorrect usage here.

However for loaded images it's more complicated of course.. The two options should definitely exist as enum variants so they can be set by decoders. The whole decoder path is a separate issue report though. I don't expect to fix any of that in this PR. We should set up the loading path mostly correctly before thinking about the storage path. Also I would argue that the current state, omitting a profile when one was present, is the same error as writing an incorrect one except not only present in files which makes me more easy about potential incorrectness for the unknown case as well.

There is one thing to worry about though, that is copy operations from buffer to buffer. `copy_from_color_space` is explicitly document as preserving each buffer's attributes but for other copies it's not entirely clear what semantics _should_ be. I'll make another pass over the buffers for that.

197g commented:
Supporting color spaces with edit operations and `into_` ballooned the code size by ~800 lines but that should be easily worth the consistency. The only tragedy is that we can not do it in `DynamicImage::to`. The output pixel type is _too broad_ so it can not be avoided that it calls either `FromColor<Luma<_>>` or `FromColor<Rgba<_>>` incorrectly. There is just no mechanism available to detect the type. Not even `TypeId` since it is not constrained as `'static`.

197g commented:
Cleaned up documentation, added the bells-and-whistles to get luma/rgb conversion with all primaries (almost never need the fallback) @fintelia: Final review time?

197g commented on src/images/buffer.rs:
```diff
@@ -1406,6 +1443,167 @@ where
+pub struct ConvertColorOptions {
+    /// A pre-calculated transform. This is only used when the actual colors of the input and
+    /// output image match the color spaces with which the was constructed.
+    pub transform: Option<CicpTransform>,
```
I've marked the option struct `!Send + !Sync` by a `PhantomData` attribute to ensure that even local options can be added. Except if they have lifetimes, but if we don't tie that to the lifetime of `self` in all the calls its usefulness would always be limited. In any case with `!Send + !Sync` we could add an implementation-detail `Rc<RefCell<_>>` in the worst case. I think we'll be limited by available throughput for quite a while before we can actually explore the implications.

This should be as flexible as we can make it without impacting usability of values.

197g replied:
It's rather necessary for performance, so I'll currently hide the field as private which makes it useful in our tests. (The full coverage table will take far longer if we were to recompute the luts each time).

Shnatsel commented:
I'm not sure I understand this PR, so I don't feel qualified to update the changelog in #2555, but I feel this should be included.

Also note that this pulls in `chrono` which in turn pulls in a bunch of transitive dependencies:

      Adding android-tzdata v0.1.1
      Adding android_system_properties v0.1.5
      Adding chrono v0.4.41
      Adding core-foundation-sys v0.8.7
      Adding iana-time-zone v0.1.63
      Adding iana-time-zone-haiku v0.1.2
      Adding js-sys v0.3.77
      Adding windows-core v0.61.2
      Adding windows-implement v0.60.0
      Adding windows-interface v0.59.1
      Adding windows-link v0.1.3
      Adding windows-result v0.3.4
      Adding windows-strings v0.4.2

Many of these are platform-specific, so nobody gets all of this at once, but the battle with dependency sprawl is lost by a thousand cuts. I've opened https://github.com/awxkee/moxcms/issues/97 to track this upstream.

197g commented:
#97 is already merged upstream :tada: With regards to the feature, can you try again with `0.7.5`? I believe the slowdown is almost fully related to iana timezones in chrono. It would be surprising to me if the pure _code_ was responsible. We surely want the primaries and luma coefficients even if we were to make the `Transform` part optional; but those also come from moxcms. If there is still significant slowdown then we might split that crate into a much more minimal `cicp` (the crate name is free).

Shnatsel commented on src/images/dynimage.rs:
```diff
@@ -43,6 +45,27 @@ use crate::{
+/// linearize colors internally as you might be used to from GPU shader programming. Their return
+/// values however copy the color space annotation of the source.
+///
+/// The IO functions do _not yet_ write ICC or CICP indications into the result formats. We're
+/// aware of this problem, it is tracked in [#2493] and [#1460].
```
Ah, I see. Could you open a new issue about that? I don't think the existing ones explain that adequately.
